go/vt/vtgate/planbuilder/route_planning.go

```diff
@@ -152,6 +147,80 @@ func optimizeSubQuery(ctx planningContext, op *abstract.SubQuery) (queryTree, er
 	return outerTree, nil
 }
 
+func tryMergeSubQuery(ctx planningContext, outer, subq queryTree, subQueryInner *abstract.SubQueryInner, joinPredicates []sqlparser.Expr, merger mergeFunc) (queryTree, error) {
+	var merged queryTree
+	var err error
+	switch outerTree := outer.(type) {
+	case *routeTree:
+		merged, err = tryMerge(ctx, outerTree, subq, joinPredicates, merger)
+		if err != nil {
+			return nil, err
+		}
+		return merged, err
+	case *joinTree:
+		if outerTree.outer {
+			return nil, nil
+		}
+		newMergefunc := func(a, b *routeTree) (*routeTree, error) {
+			rt, err := merger(a, b)
+			if err != nil {
+				return nil, err
+			}
+			otherTree := outerTree.lhs
+			if a.tableID() == outerTree.lhs.tableID() {
+				otherTree = outerTree.rhs
+			}
+
+			var rewriteError error
+			sqlparser.Rewrite(subQueryInner.SelectStatement.Where.Expr, func(cursor *sqlparser.Cursor) bool {
+				sqlNode := cursor.Node()
+				switch node := sqlNode.(type) {
+				case *sqlparser.ColName:
+					if ctx.semTable.Dependencies(node).IsSolvedBy(otherTree.tableID()) {
+						bindVar := node.CompliantName()
+						cursor.Replace(sqlparser.NewArgument(bindVar))
+						_, alreadyExists := outerTree.vars[bindVar]
+						if alreadyExists {
+							return false
+						}
+						columnIndexes, err := otherTree.pushOutputColumns([]*sqlparser.ColName{node}, ctx.semTable)
+						if err != nil {
+							rewriteError = err
+							return false
+						}
+						columnIndex := columnIndexes[0]
+						outerTree.vars[bindVar] = columnIndex
+						return false
+					}
+				}
+				return true
+			}, nil)
+
+			return rt, rewriteError
```

Comment:
We should move this to a non-exported method like we did for `mergeSubQuery`

Reply:
done